packages/dds/map/src/mapKernel.ts

```diff
@@ -796,12 +1198,46 @@ export class MapKernel {
 			this.pendingKeys.delete(op.key);
 		}
 
+		/**
+		 * Considering a new message id will be assigned, if the op is of type set, it is necessary to replace
+		 * all "old information" with new message id
+		 */
+		if (op.type === "set") {
+			const pendingSetIds = this.pendingSetTracker.get(op.key);
+			if (pendingSetIds !== undefined) {
+				const topPendingSetId = pendingSetIds?.shift() as number;
+				if (pendingSetIds?.length === 0) {
+					this.pendingSetTracker.delete(op.key);
+				}
+				if (this.localKeysIndexTracker.has(topPendingSetId)) {
+					this.localKeysIndexTracker.delete(topPendingSetId);
+					if (pendingSetIds.length > 0) {
+						this.localKeysIndexTracker.set(op.key, pendingSetIds[0]);
+					}
+				}
+			}
+		}
+
 		// We don't reuse the metadata pendingMessageId but send a new one on each submit.
 		const pendingMessageId = this.getMapKeyMessageId(op);
+		this.updatePendingSetIds(op, pendingMessageId);
+
 		const localMetadata =
 			localOpMetadata.type === "edit"
-				? { type: "edit", pendingMessageId, previousValue: localOpMetadata.previousValue }
-				: { type: "add", pendingMessageId };
+				? {
+						type: "edit",
+						pendingMessageId,
+						previousValue: localOpMetadata.previousValue,
+				  }
+				: localOpMetadata.type === "add"
+				? { type: "add", pendingMessageId }
+				: {
+						type: "delete",
+						pendingMessageId,
+						previousValue: localOpMetadata.previousValue,
+						previousIndex: localOpMetadata.previousIndex,
+				  };
```

Comment:
not undefined was already checked above
```suggestion
				if (pendingSetIds.length === 0) {
```